optimizer/lib/transformers/AddBlurryImagePlaceholders.js

```diff
@@ -34,24 +40,39 @@ class AddBlurryImagePlaceholders {
     const html = tree.root.firstChildByTag('html');
     const body = html.firstChildByTag('body');
     const promises = [];
+    let currPlaceholderCount = 0;
     for (let node = body; node !== null; node = node.nextNode()) {
       const {tagName} = node;
       let src;
+      if (tagName === 'template') {
+        node = nextNode(node);
+        continue;
+      }
       if (tagName === 'amp-img') {
         src = node.attribs.src;
       }
       if (tagName === 'amp-video' && node.attribs.poster) {
         src = node.attribs.poster;
       }
-      if (src && !this.hasPlaceholder_(node)) {
-        promises.push(this.addBlurryPlaceholder_(tree, src).then((imgChild) => {
-          node.appendChild(imgChild);
-        }));
+      if (currPlaceholderCount >= MAX_BLURRED_PLACEHOLDERS) {
+        break;
+      }
+      if (this.shouldAddBlurryPlaceholder_(node, src, tagName)) {
+        promises.push(this.addBlurryPlaceholder_(tree, src)
+          .then((imgChild) => {
+            if (currPlaceholderCount >= MAX_BLURRED_PLACEHOLDERS) {
+              return;
+            }
```

Comment:
Why check this twice?

Reply:
When I don't have this second check, then the adds_max_amount_of_blurs test fails because of the synchronicity of the node traversal if i understand it correctly. The test is supposed to stop adding blurs after the first five, but it adds one to the sixth element as well.

So if I only have the first "placeholder amount" check, then >5 blurs can be added because  the placeholder amount is only increased after a placeholder is added, and multiple blurs can be in the process of being added at once. 

If you think my test is failing for another reason that I'm not understanding correctly or if you want me to redo the code, then please let me know! Just wanted to explain my logic.

Reply:
Try moving the `currPlaceholderCount++` line right below the first check.